OIDC Provider Module (#5616)
mendix/docs: content/en/docs/appstore/modules/oidc-provider.md

---
title: "OIDC Provider"
url: /appstore/modules/oidc-provider/
category: "Modules"
description: "Describes the configuration and usage of the OIDC Provider module, which is available in the Mendix Marketplace."
tags: ["marketplace", "marketplace component", "OIDC", "IdP", "identity provider", "platform support", "Provider"]
#If moving or renaming this doc file, implement a temporary redirect and let the respective team know they should update the URL in the product. See Mapping to Products for more details.
# Linked from https://marketplace.mendix.com/link/component/+++++++
---

## 1 Introduction

> {{% todo %}}Add link to module in Marketplace{{% /todo %}}

Author: Need to add a link to the module.

Reviewer: @MarkvanMents As it is not yet published on Marketplace, I will add the comment here once I publish the module

Reviewer: Hi @MarkvanMents 
OIDC Provider module link https://marketplace.mendix.com/link/component/210202

The [OIDC Provider](https://example.com) can be used to build a Mendix app that acts as an Identity provider(IdP) for other apps. This app then acts as an IAM broker, providing a Single Sign-On (SSO) experience for the end-users of those applications.

The module supports responsive browser-based applications and has been tested with applications that use the OIDC SSO module. This module can be used in Mendix version 9.12.5 and above.

The idea is that you set up a single Mendix app which uses the [OIDC SSO](https://marketplace.mendix.com/link/component/120371) module to authenticate end-users with your central IdP. The same app also acts as an OIDC client for your other apps to use as the IdP for OIDC SSO. THis means it is working as an IAM (Identity and Access Management) broker. You can easily add or remove apps from the IAM Broker app within the Mendix ecosystem using an API without each app and relevant user roles having to be added to your central IdP. However, you retain all the benefits of your central IdP in controlling on- and offboarding of users.

> {{< figure src="/attachments/appstore/modules/oidc-provider/typical-usage.png" >}}

Author: This picture does not reflect what is being described in this document. I got very confused thinking we had a back-end IdP whereas we are just adding users through the administration module.

Reviewer: Hi @MarkvanMents ,
Now in the latest version of the module, we will add users from OIDC provider module. Added users have connection with Admin module in the module itself

Author: What I mean is that the picture shows that the users which are exposed by the OIDC Provider module are retrieved from the Schwarz IdP - not entered manually in the Administration module.
@JaapF, what do you think?

Author: Also, should this image be more generic and not specifically mention **Schwarz IDP**?

Reviewer 2: @Mark, I clarified this to Sharisma this morning as well. Users can be created in 2 ways in the broker:
- via JIT user provisioning (IdP --> OIDC Client)
- via Administration module

We should clarify that both are valid options. I consider JIT as the most viable approach, but that's up to the customer.

Reviewer 2: Yes, the picture should obviosuly not have Schwarz. I updated the picture in the dropbox paper
https://paper.dropbox.com/doc/OIDC-Provider--By2uC5lONCl20SMsodVsF3~YAg-jLCM4Hz40c2sIHBNnqaTo

Author: @JaapF 
I'll have a look and see if we can make these use cases clearer.

In the JIT situation, where are the user credentials shared from? It doesn't make sense to keep them in persistable entities in the database, but should we give the user advice in terms of when they need to refresh the credentials from the corporate IdP. For example, it doesn't make sense to take the user information from the corporate IdP and never go back to check that the user is still active, or that their roles haven't changed. Do we expect to do that with every authentication request (i.e. we NEVER store any user information in the app) or is that a potential performance drain. And how should customers handle users who still have roles in the OIDC Provider module, but are no longer in the corporate IdP?

### 3.1 Configure Security

To configure the app security, do the following:

1. Open the [User Roles](/refguide/user-roles/) in the app and **Edit** the **Administrator** user role.
1. **Edit** the **Module roles**
1. Assign the following module roles to the Administrator:
    1. MxModelReflection
        * ModelAdministrator
        * Readonly
        * TokenUser
    1. OpenIDConnectProvider
        * Admin

### 3.2 Configure App Modules

You now need to configure the OIDC Provider and Administration modules in the IAM Broker app.

1. Open the **Domain model** of the **Administration** module.
1. Edit the **Account** entity.
    1. Switch to the **Access Rules** tab.
    1. **Edit** the **Administrator** module role.
    1. Give **Read** **Access rights** to the **AccountDetail_Account** association.
1. Open the **Runtime** tab of the app **Settings**.
1. Add the **OpenIDConnectProvider.ASU_Start_All** microflow as an [After Startup](/refguide/app-settings/#after-startup) microflow.
1. Finally, ensure that your app has the following Administration navigation items:

    | Caption | Action | User Roles |
    | --- | --- | --- |

>     | Accounts | Open page 'Administration.AccountDetail_Overview' | Administrator, UserManager |

Reviewer: Can you please change this?
OpenIDConnectProvider.AccountDetail_Overview

and remove Usermanager userrole from all navigation items

Author: Done

    | Active Sessions | Open page 'Administration.ActiveSessions' | Administrator, UserManager |
    | Scheduled Events | Open page 'Administration.ScheduledEvents' | Administrator, UserManager |
    | Runtime Instances | Open page 'Administration.RunteimInstances' | Administrator, UserManager |
    | My Account | Call microflow 'Administration.ManageMyAccount' | User |
    | OpenID Connect | Open page 'OpenIDConnectProvider.OpenIDConnect_Dashboard' | Administrator |

### 3.3 Configure OIDC Provider

The rest of the configuration can be performed through the app.

1. Deploy and run the app.
1. Login as an administrator.
1. Use the **Accounts** navigation item.

> 1. Click **New local user** and create the user account(s) you need in your IAM Broker — for example, a new active account with the **User role** set to **User**.

Author: Does this mean that the IAM Broker app has to have all the roles defined which are used in the apps which are using it. So if I have an app which has "teacher" and "student" roles, these roles have to be defined in the IAM Broker app so they can be propagated to the OIDC Client?

Should we mention this somewhere?

Reviewer: We will create an account with a userrole in OIDCProvider to login with the credentials via OIDC Client.

If we don't give any userrole when we create users in OIDC Provider, we are unable to login to OIDC Client using the IDP OIDC Provider
<img width="396" alt="image" src="https://user-images.githubusercontent.com/110099605/218022171-50c89034-79e8-43ba-9835-32a6ced044d8.png">

When we login via OIDC Client app by using specific IDP, by default the user will get 'User' userrole in OIDC CLient after successful login

Author: So, if I understand it, the user roles do not get propagated from the OIDC Provider. You can only login as "User", not as any other role (such as "teacher" or "student" or "Administrator").

Is that a big restriction given the way that Mendix security works? The OIDC SSO module even has a microflow to pick up roles for you: `OIDC.ACT_Token_CustomATPRetrieveRoles`

Reviewer: Hi @MarkvanMents ,
Any user we are creating in the OIDC Provider should compulsorily have a user role(ex:user,teacher,student or Admin). When we try login from OIDC Client by clicking on "Login via OIDC Provider", if login is successful, user in the OIDC Client gets "User" userrole by default.

If customer wants to use OIDC.ACT_Token_CustomATPRetrieveRoles token to get user roles from access token and assign them to user, then the user roles included in the access token needs to be present in OIDC Client as well to map them to user

Reviewer 2: If a customer decides to connect 100 apps to the IAM broker and each of those apps have 2 roles, this would result in 200 userroles.  These could be made known in the IAM broker but not necesarily. In either case the OIDC Provoder module doesn't know/care. We do not the capability in the OIDCProvider module to include 'custom claims' in an ID-token. This could be aatributes like roles,groups,entitelements,userroles, whatever.
@Mark - I hope this clarifies. We should focus on documenting the OIDC provider module, without describing how to build a IAM Broker (which could have some kind of authorization logic).
@Sharishma - this also explains why I think it is important to have custom claims. I think currently a customer would have to customize the OIDC Provider module to make custom claims a possibility.

Author: So if the user has several roles (e.g. "teacher" or "student") in the app containing the OIDC Provider app, then these roles will be added to the access token from the OIDC Provider?

Reviewer 2: @MarkvanMents to the provider-app "teacher" and "student" is just data. Those are not roles in the app containing the provider, they would be roles in the target app.

1. Select **Automatic Registration**.
1. Take a copy of the **Bearer Registration Access token** and the **Registration URI**.
1. In your API testing tool, set up the following:
    * a **POST** command to the **Registration URI** with the endpoint `/oidc/register`, for example, for a locally-deployed app, `http://localhost:8080/oidc/register`
    * an HTTP header with **Key** = "Authorization" and **Value** = **Bearer Registration Access token**
1. Send the command. If it is working, you should get a response which has a body which resembles the example below:

    ```json {linenos=false}
    {
    "post_logout_redirect_uris": [
    "http://localhost:8081/logout"
    ],
    "grant_types": [
    "authorization_code",
    "password"
    ],
    "client_secret_expires_at": 0,
    "scope": "openid",
    "client_secret": "c46591bd-8fae-4f90-9efb-b5973bea04df",
    "redirect_uris": [
    "http://localhost:8081/oauth/v2/callback"
    ],
    "client_id_issued_at": 1675940602,
    "backchannel_logout_uri": "http://localhost:8081/logout",
    "client_name": "DemoClient1",
    "client_id": "DemoClient1"
    }

>     ```

Author: Should we have all these instructions on how to test it. It isn't very low-code?
Should we perhaps describe it as an API, like we do in https://docs.mendix.com/apidocs-mxsdk/apidocs/deploy-api/#list-apps?

Reviewer: @MarkvanMents If they want to create a DemoClient, they can use this URI

Author: I think we need more explanation as to what we mean by a DemoClient. Most users will want to create real clients. See comment below on what I think needs more explanation 👇

#### 3.3.2 Manual Client Registration
    
If you cannot use automatic registration, you can register the client manually.

1. Select **Manual Registration**.
1. Add the following information:
    * **Client Name**
    * **Client ID**
    * **Alias**
    * **Client Secret** 
    * **Post Logout redirect URI** – for example, for testing a local OIDC Provider app on port `8081`, `http://localhost:8081/logout` 
    * **Redirect URI** – for example, for testing a local OIDC Provider app on port `8081`, `http://localhost:8081/oauth/v2/callback`
    * **Back channel logout session support**
    * **Front channel Logout URI**

>     * **Requires PKCE** – this needs to be enabled in both the client and the server to use PKCE

Author: Can we explain these a bit better here. For example, what is the difference between `Client Name` and `Client ID`. Do they both have to be unique?

Reviewer: Hi @MarkvanMents ,
According to my knowledge they can mention anything as client name. For example: application name
Client Name need not to be unique
but ClientID needs to be unique

Reviewer 2: Actually I think clientID is meant to be a unique unmutable identifier whereas ClientName is a human-readable descriptive identifier for a client.
That's how it should be.
In practice client-id's are often made readable for humans as well. They must be unique and unmutable at least. ClientName should also be unique, but apparantly the provider module does not enforce that, right @SharishmaPodduturi999 ?

Author: Done https://github.com/mendix/docs/pull/5616/commits/6b4c1384c138cd39b637eda5db44a88ff7dd2d99

* Mendix Solution Vendors (MSVs) who want to simplify the microservice architecture of their solution from the customer by using the IAM Broker as a single IAM integration point for their customers.

### 1.2 Features and Limitations

The OIDC Provider has the following features and limitations:

**Features**

* It supports responsive web applications, using the common OAuth Authorization Code grant.
* Your apps can be registered as an OIDC client with the OIDC Provider using the client registration API or client configuration screen.
* It publishes a well-known endpoint to communicate endpoints and other IdP characteristics to client applications. This can be used by clients to retrieve the keys needed to validate ID-tokens that they receive.
* It supports the OIDC ‘nonce’ parameter, PKCE, and multiple client authentication methods ( client_secret_post, client_secret_basic) as security features.

>     

Reviewer 2: Hi @MarkvanMents 
In the original dropbox paper we had a sentence that said:
"Supports front channel and back-channel logout as public beta features."

I don't know/remember why it's not included in this PR.
I do want to stress that the logout functionality is not even beta but alpha: it is not yet tested by the developer and it is probably not feature complete.

Reviewer 2: Maybe we should say in limitations: "The front channel and back-channel logout are available as ‘alpha’ features only."

Reviewer: @JaapF now front channel and back channel options are available on UI of OID provider. Should we remove them from UI or shall we put them as they are?

Author: @JaapF please let me know what needs to be put in the documentation here, after your conversation with Sharishma.

Reviewer 2: Let's keep it. We tell customers it's a alpha implementation.

Author: I've added this in the suggestion in the limitations.

**Limitations**

* The hybrid and client credential grants are not supported, although the OIDC Provider may contain some (rudimentary) implementation to support them.

> 

Reviewer 2: We have an additional limitation:

"The OIDC Provider module features an initial rudimentary implementation of a the possibility to set custom claims in the ID-tokens that are generated. This feature is not yet working properly as the names of a a custom attroibute that you would define is not properly set in the ID-token. If you want to set custom claims, you would have to customise the OIDc Provider module"

Reviewer 2: @SharishmaPodduturi999 plesae review.
Also I think we need to provide clarity on the non-custom claims that the OIDC Provider module generates.
Maybe we can include an example of both an ID-topken and an Access Token that the OIDC Provider generates and sends to the app.

Reviewer: @JaapF @MarkvanMents 
**Non-custom claims of Access Token:**
{
  "aud": "DemoClient",
  "sub": "3",
  "nbf": 1681969726,
  "scope": "",
  "iss": "http://localhost:8080/",
  "name": "test",
  "exp": 1682056126,
  "iat": 1681969726
}

**Non-custom claims of ID Token**
{
  "com.mendix.user.language": "en_US",
  "sub": "test",
  "iss": "http://localhost:8080/",
  "com.mendix.user.entity": "Administration.Account",
  "nonce": "k5CDLkTE7Q61Q0cUTSgy",
  "com.mendix.user.attributes": {
    "Email": "",
    "FullName": "test",
    "IsLocalUser": "true"
  },
  "aud": "DemoClient",
  "scope": "",
  "exp": 1681970318,
  "com.mendix.user.timezone": "",
  "iat": 1681969718,
  "com.mendix.user.roles": [
    "User"
  ]
}

Author: @JaapF 
I've added the extra limitation in this commit. https://github.com/mendix/docs/pull/5616/commits/5bba674e8a2b0a7a2aeb1f5dbb6440815e99f53f
If we have these limitations, are they serious enough to mark this release as a Beta?

I've added the claim formats to a new section at the end of the document in the same commit.

### 1.3 Dependencies

This module requires your app to be using Mendix 9.12.5+ or higher
The following modules need to be imported into your app

* [Community Commons](https://marketplace.mendix.com/link/component/170) – see [Community Commons](/appstore/modules/community-commons-function-library/) documentation
* [Mx Model reflection](https://marketplace.mendix.com/link/component/69) – see [Mx Model Reflection](/appstore/modules/model-reflection/) documentation

## 2 Installation

To install the OIDC Provider module in your IAM broker app, you need to import the following into your app:

* [OIDC Provider](https://example.com)
* [Community Commons](https://marketplace.mendix.com/link/component/170)
* [Mx Model reflection](https://marketplace.mendix.com/link/component/69)

> * [Administration](https://marketplace.mendix.com/link/component/23513)

Reviewer: @MarkvanMents Can you remove Administration module here?

Author: This is linked to my comment above. We need to advise customers how to deal with data which is coming from their IdP (store it somewhere, always pull the latest version then forward it and delete it). In the examples in the documentation). If they need to store user information somewhere, then they will need the Administration module.

Author: ?

Reviewer: @MarkvanMents 
Administration is a default module in Marketplace when anyone creates a new application.So I am thinking to remove it from dependencies section.
agree?

Author: I would leave it here if the OIDC Provider module depends on it. If customers have their own implementation of user administration they could, theoretically, remove it.

            "post_logout_redirect_uris" : [http://localhost:8081/logout],
            "grant_types": [ "authorization_code","password" ],
            "scope": "openid"
        }
        ```

1. Send the command. If it is working, you should get a response which has a body which resembles the example below:

    ```json {linenos=false}
    {
    "post_logout_redirect_uris": [
    "http://localhost:8081/logout"
    ],
    "grant_types": [
    "authorization_code",
    "password"
    ],
    "client_secret_expires_at": 0,
    "scope": "openid",
    "client_secret": "ClientSecret",
    "redirect_uris": [
    "http://localhost:8081/oauth/v2/callback"
    ],
    "client_id_issued_at": 1675940602,
    "backchannel_logout_uri": "http://localhost:8081/logout",
    "client_name": "ClientName",
    "client_id": "ClientID"
    }
    ```

#### 3.3.2 Manual Client Registration
    
If you cannot use automatic registration, you can register the client manually.

1. Select **Manual Registration**.
1. Add the following information:
    * **Client Name** – a name for this client so that it is easy to identify
    * **Client ID** – a unique string which identifies this client

>     * **Alias**

Author: What should go in Alias. Is it different from Client Name?

Reviewer 2: @SharishmaPodduturi999 can you answer. I don't know. "Alias" is not defined by OAuth / OIDC specs - it's proprietary. We may as well remove from implementation as far as I'm concerned.

Reviewer: @JaapF @MarkvanMents 
'Alias' can be same as ClientName or can be different also

Alias is used in 7 pages as of now. Shall we plan it remove in future release

Author: Added in this commit: https://github.com/mendix/docs/pull/5616/commits/ab80941f505533587620f6a41c3bf3ba4bf3aab4

>     * **Client Secret** – a secure string to allow the client to authenticate to the OIDC Provider module

Author: Is this a good description?

Reviewer 2: We might as well call it a 'client password' - I guess that's more clear.

Author: Changed in https://github.com/mendix/docs/pull/5616/commits/ab80941f505533587620f6a41c3bf3ba4bf3aab4

---
title: "OIDC Provider"
url: /appstore/modules/oidc-provider/
category: "Modules"
description: "Describes the configuration and usage of the OIDC Provider module, which is available in the Mendix Marketplace."
tags: ["marketplace", "marketplace component", "OIDC", "IdP", "identity provider", "platform support", "Provider"]
#If moving or renaming this doc file, implement a temporary redirect and let the respective team know they should update the URL in the product. See Mapping to Products for more details.
# Linked from https://marketplace.mendix.com/link/component/+++++++
---

## 1 Introduction

{{% todo %}}Add link to module in Marketplace{{% /todo %}}

> The [OIDC Provider](https://example.com) can be used to build a Mendix app that acts as an Identity provider(IdP) for other apps. This app then acts as an IAM broker, providing a Single Sign-On (SSO) experience for the end-users of those applications.

Reviewer 2: An app using OIDC Provider acts as a broker when you combine it with a OIDC client, as shown in the picture https://paper.dropbox.com/doc/OIDC-Provider--By2uC5lONCl20SMsodVsF3~YAg-jLCM4Hz40c2sIHBNnqaTo.
So that is the main usage scenario.

Author: @JaapF, I wonder if we should explain that usage scenario better, rather than describing how we add things manually? For example, showing where the OIDC SSO module records the user from the <Schwartz> IdP and how that then gets shared via the OIDC Provider.

Or will customers be able to work this out for themselves?

Author: We still don't really describe how to get information from a back-end IdP and process it in the OIDC Provider. Should we give advice about this, for example
* don't cache anything locally - always go back to the IdP source
* only cache information for a certain time - then go back to the IdP source
* cache information for the duration of a session - then delete and go back to the IdP source

Perhaps a simple example would help?

Reviewer 2: @MarkvanMents
 "This app then acts as an IAM Broker" --> "In OIDC terminilogy: the app acts as an OpenID Provider. If this app delegates authentication of endusers to an other IdP, it acts as a IAM Broker."

Author: @JaapF is this OK?
```suggestion
The [OIDC Provider](https://example.com) can be used to build a Mendix app that acts as an OpenID provider for other apps, providing a Single Sign-On (SSO) experience for the end-users of those applications. This app could also delegate authentication of end-users to another Identity provider (IdP), causing it to act as an IAM broker.
```

Reviewer 2: 👍

The module supports responsive browser-based applications and has been tested with applications that use the OIDC SSO module. This module can be used in Mendix version 9.12.5 and above.

> The idea is that you set up a single Mendix app which uses the [OIDC SSO](https://marketplace.mendix.com/link/component/120371) module to authenticate end-users with your central IdP. The same app also acts as an OIDC client for your other apps to use as the IdP for OIDC SSO. THis means it is working as an IAM (Identity and Access Management) broker. You can easily add or remove apps from the IAM Broker app within the Mendix ecosystem using an API without each app and relevant user roles having to be added to your central IdP. However, you retain all the benefits of your central IdP in controlling on- and offboarding of users.

Reviewer 2: In think you mean to say
"..same app acts as an OIDC Provider ..." not client.

Author: Done https://github.com/mendix/docs/pull/5616/commits/6b4c1384c138cd39b637eda5db44a88ff7dd2d99

{{< figure src="/attachments/appstore/modules/oidc-provider/typical-usage.png" >}}

### 1.1 Typical Usage Scenarios

The following are usage scenarios that would be achievable with the OIDC Broker.

* Mendix customers that want to build an IAM Broker solution that would hide the complexity of a multitude of Mendix apps from their corporate IdP.  By having those apps delegate authentication to the broker and have the broker delegate authentication to their IdP, only one OAuth client needs to be configured at their IdP.  A deployment pipeline can register additional Mendix apps with the IAM Broker in an automated fashion via an API. 

> * Mendix Solution Vendors (MSVs) who want to simplify the microservice architecture of their solution from the customer by using the IAM Broker as a single IAM integration point for their customers.

Reviewer 2: In my draft it said "In a similar fashion, Mendix Solution Vendors (MSVs) may want to hide the microservice architecture of their solution from the customer by using the IAM Broker as a single IAM integration point for their customers."
No it says "...simplify ..architecture ... from the customer"

The complexity of the microservice architecture would not be simplified but hidden for their customers.

Reviewer 2: I'm reading my own sentence back - not clear. Should have been:

"Mendix Solution Vendors (MSVs) may want to hide the microservice architecture of their solution from the customer. When building an IAM Broker they can offer their customers a single SSO integration point."

Author: @JaapF is this OK?
```suggestion
* Mendix Solution Vendors (MSVs) may want to hide the microservice architecture of their solution from the customer. By building an IAM Broker they can offer their customers a single SSO integration point.
```

Reviewer 2: 👍

### 1.2 Features and Limitations

The OIDC Provider has the following features and limitations:

**Features**

> 

Reviewer 2: We jump straight into the OIDC-technicalities.
Maybe we should have a simpler feature list for low-code developers that don't know about this stuff.

For that developer we could say:
- works/integrates with OIDC SSO module
- allows for registration of Mendix apps via the client registration endpoint.

Author: @JaapF is this OK?
```suggestion

* allows for registration of Mendix apps via the client registration endpoint
* works/integrates with the platform-supported [OIDC SSO](/appstore/modules/oidc/) module
```

Reviewer 2: ok

* It supports responsive web applications, using the common OAuth Authorization Code grant.
* Your apps can be registered as an OIDC client with the OIDC Provider using the client registration API or client configuration screen.

> * It publishes a well-known endpoint to communicate endpoints and other IdP characteristics to client applications. This can be used by clients to retrieve the keys needed to validate ID-tokens that they receive.

Reviewer 2: 2nd sentence improvement
"Mendix apps apps using OIDC SSO will consume this endpoint, e.g. to retrieve...."

Reviewer 2: In the dropbox paper, I added an endpoint/integration overview.
We could include that, I'm not sure if we should.

Author: @JaapF is this OK?
I feel we are making the documentation less generic - there may be customers out there who want to use the well-known endpoint from something other than the OIDC SSO module?
```suggestion
* It publishes a well-known endpoint to communicate endpoints and other IdP characteristics to client applications. Mendix apps using the OIDC SSO module will consume this endpoint to perform actions like retrieving the keys needed to validate ID-tokens that they receive.
```

Reviewer 2: Hi Mark, I think it is good as it is.
I think it is still generic and at the same time we're saying that the OID SSO moidule is following the generic principles.

* It supports the OIDC ‘nonce’ parameter, PKCE, and multiple client authentication methods ( client_secret_post, client_secret_basic) as security features.
    
**Limitations**

* The hybrid and client credential grants are not supported, although the OIDC Provider may contain some (rudimentary) implementation to support them.

### 1.3 Dependencies

This module requires your app to be using Mendix 9.12.5+ or higher
The following modules need to be imported into your app

* [Community Commons](https://marketplace.mendix.com/link/component/170) – see [Community Commons](/appstore/modules/community-commons-function-library/) documentation
* [Mx Model reflection](https://marketplace.mendix.com/link/component/69) – see [Mx Model Reflection](/appstore/modules/model-reflection/) documentation

> * [Administration](https://marketplace.mendix.com/link/component/23513) – see [Administration](/appstore/modules/administration/) documentation

Reviewer: Hi @MarkvanMents Can you please remove Administration module as a dependency as it is a default marketplace module when we create a new project

Author: Done https://github.com/mendix/docs/pull/5616/commits/6b4c1384c138cd39b637eda5db44a88ff7dd2d99

## 2 Installation

To install the OIDC Provider module in your IAM broker app, you need to import the following into your app:

* [OIDC Provider](https://example.com)
* [Community Commons](https://marketplace.mendix.com/link/component/170)
* [Mx Model reflection](https://marketplace.mendix.com/link/component/69)
* [Administration](https://marketplace.mendix.com/link/component/23513)

The module then needs to be configured as described below.

## 3 Configuring IAM Broker

The steps in this section set up the app which is running as an IAM Broker using the OIDC Provider module.

### 3.1 Configure Security

To configure the app security, do the following:

1. Open the [User Roles](/refguide/user-roles/) in the app and **Edit** the **Administrator** user role.
1. **Edit** the **Module roles**
1. Assign the following module roles to the Administrator:
    1. MxModelReflection
        * ModelAdministrator
        * Readonly
        * TokenUser
    1. OpenIDConnectProvider
        * Admin

### 3.2 Configure App Modules

> You now need to configure the OIDC Provider and Administration modules in the IAM Broker app.

Reviewer 2: As a reader I would like to understand why?
I think th answer is: because user account records (Admin module) are linked/replicated into AccountDetails (OIDC Provider).
If my understanding is correct i would suggest.

"Every user that is known in Admin module, also needs to be known in the OIDC Provider module. Technically this means OIDCProvier.AccountDetail records are created for every Administration.Account record. This can be achieved by the following steps:"

@SharishmaPodduturi999 and @MarkvanMents plesae review if this makes sense to you and is correct.

Reviewer: Yes @JaapF You are correct.it makes sense to me

but a small addition in the sentence
Every user that is known in Admin module, also needs to be known in the OIDC Provider module. Technically this means OIDCProvier.AccountDetail records are created for every Administration.Account record `in the time of login`. This can be achieved by the following steps

Reviewer 2: 👍

Author: @JaapF - does this look OK?
```suggestion
Every end-user that is known in Admin module also needs to be known in the OIDC Provider module. Access has to be given to allow an OIDCProvier.AccountDetail record to be created for every Administration.Account record when the end-user signs in. This can be achieved by the following steps:
```

Reviewer 2: I think it is ok - I'm relying a bit on Sharishma's review here.

Reviewer: It looks good to me @JaapF

    1. **Edit** the **Administrator** module role.
    1. Give **Read** **Access rights** to the **AccountDetail_Account** association.
1. Open the **Runtime** tab of the app **Settings**.
1. Add the **OpenIDConnectProvider.ASU_Start_All** microflow as an [After Startup](/refguide/app-settings/#after-startup) microflow.
1. Finally, ensure that your app has the following Administration navigation items:

    | Caption | Action | User Roles |
    | --- | --- | --- |
    | Accounts | Open page 'OpenIDConnectProvider.AccountDetail_Overview' | Administrator |
    | Active Sessions | Open page 'Administration.ActiveSessions' | Administrator |
    | Scheduled Events | Open page 'Administration.ScheduledEvents' | Administrator |
    | Runtime Instances | Open page 'Administration.RunteimInstances' | Administrator |
    | My Account | Call microflow 'Administration.ManageMyAccount' | User |
    | OpenID Connect | Open page 'OpenIDConnectProvider.OpenIDConnect_Dashboard' | Administrator |

### 3.3 Configure OIDC Provider

The rest of the configuration can be performed through the app.

1. Deploy and run the app.
1. Login as an administrator.
1. Use the **Accounts** navigation item.

> 1. Click **New local user** and create the user account(s) you need in your IAM Broker — for example, a new active account with the **User role** set to **User**.

Author: **Comment from @JaapF 6 months ago in original Paper document** 
> I would expect that if customer uses OIDC Provider module to build a Private IAM Brokjer (like Schwarz is doing), users are net necessarily created via Administration module but could be create through a login (Just-in-Time provisioning).

Reviewer 2: Yes, see my previous comment. Users can either be created via OIDC-SSO JIT-User-provisioning or via Administrator module.

Author: Discussed above. Resolved.

> 1. Click **New local user** and create the user account(s) you need in your IAM Broker — for example, a new active account with the **User role** set to **User**.

Reviewer 2: Let's say "in your OIDC Provider app". I see a risk of having to document how to build a broker. We're docuemnting only the provider.

> 1. Click **New local user** and create the user account(s) you need in your IAM Broker — for example, a new active account with the **User role** set to **User**.

Reviewer: @MarkvanMents Can you please change **New local user** to "New"

Author: Done https://github.com/mendix/docs/pull/5616/commits/6b4c1384c138cd39b637eda5db44a88ff7dd2d99

1. Use the **OpenID Connect** navigation item.
1. Open the **Server Keys** tab.
1. Click **New Key**.
1. Open the **Scopes** tab.

> 1. Create a **New** scope with the **Scope type** set to **Standard**.

Reviewer: Can you also please change this to 
Create a **New** scope with Name as openid and the **Scope type** set to **Standard**

Author: From the screenshot, I assumed that the **Name** can be chosen by the customer. Does it have to be *openid*? If so, should this be a dropdown rather than a text input field?

Reviewer: Can we give openid as an example? So that every one will have an idea

Author: I don't think _openid_ is a good example of a scope. Scopes are linked to user roles, so perhaps _User_ is a good scope for an example.

What other scope types are there - they aren't mentioned or described in the document.

Reviewer 2: In the 2nd dropobox paper, I had added some clarifications:

"Note: Make sure to not to create custom scopes that ‘collide’ with standard scopes (as per [OIDC specs](https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims)) supported by OIDC Provider:

- openid - apps will use this scope value to request identity propagation from the OIDC Provider by means of an ID-token.

We recommend to not have any userroles with the following names, as these would collide with standard scope values and may break forward compatibility with future versions of the OIDC Provider module.

- email
- profile
- phone"

@MarkvanMents , there are no other standard/default scopes supported by the OIDC Provider module.

Reviewer 2: @SharishmaPodduturi999 , do we have any logic in the provider module to handle the following scopes in a request: email, profile, phone?

According to OIDC specs:
- email --> makes the Provider include the user's email in ID-token
- phone --> makes the Provider include the user's phonenumbers in ID-token
- profile --> makes the Provider include the user's full profile in ID-token.

Reviewer: No @JaapF ,
There is no logic available to include phone and profile scopes in ID-token
but we have the logic to include email with different name 'Email' included in "com.mendix.user.attributes" claim
Example:
![image](https://user-images.githubusercontent.com/110099605/233273482-23eaaf6f-d9f1-4f8e-a677-6e73814d86b2.png)

Author: This information is now included in section 5.2 - Non-custom claims in ID-token.

1. Open the **Clients** tab.
1. Click **New Client Registration**.

>     You can register a new client in one of two ways. Follow one of the sets of instructions below:

Author: Should we make it clearer what we mean by a "client"? For example:
```suggestion
    You can register a new client (an app using the OIDC SSO module for sign in which identifies this app as its IdP) in one of two ways. Follow one of the sets of instructions below:
```

Reviewer 2: Yes Mark, I think that is a good clarification.
Full code developers would understand the cl;ient concept.
Less technical developers would not understanmd this.

#### 3.3.1 Automatic Client Registration

To check that this works, you will need a tool for testing APIs (such as [Postman](https://www.postman.com/)) which allows you to execute HTTP commands to the Registration URI of you IdP.

1. Select **Automatic Registration**.

> 1. Take a copy of the **Bearer Registration Access token** and the **Registration URI**.

Reviewer 2: I think this is the first location where we mention the RAT - how do I get one as an Admin?
@SharishmaPodduturi999 can you explain?

Reviewer: @JaapF 
These are the steps to get registration access token as admin
1. Login as Admin
2. Use the **OpenID Connect** navigation item
3. Switch to Clients tab
4. click on "New Client Registration"
5. Select **Automatic Registration**.
6. Take a copy of the **Bearer Registration Access token** and the **Registration URI**

These are already included in the document along with other steps

Reviewer 2: @SharishmaPodduturi999 , thanks yes your right. My comment was not valid - I had misunderstood: "take a copy.." is where you get the RAT. I mistakenly thought this was a step where you use the RAT, rather then obtain it.
My bad...

1. In your API testing tool, set up the following:
    * a **POST** command to the **Registration URI** with the endpoint `/oidc/register`, for example, for a locally-deployed app, `http://localhost:8080/oidc/register`
    * an HTTP header with **Key** = "Authorization" and **Value** = **Bearer Registration Access token**
    * an HTTP request body with the following format (for a Client `ClientID` which is running on host and port `localhost:8081`):

        ```json {linenos=false}
        {
            "client_id" : "ClientID",
            "client_name" : "ClientName",
            "client_secret" : "ClientSecret",
            "redirect_uris" : [ http://localhost:8081/oauth/v2/callback ],
            "backchannel_logout_uri" : http://localhost:8081/logout,
            "post_logout_redirect_uris" : [http://localhost:8081/logout],
            "grant_types": [ "authorization_code","password" ],
            "scope": "openid"
        }
        ```

1. Send the command. If it is working, you should get a response which has a body which resembles the example below:

>     ```json {linenos=false}

Author: **Comments from @JaapF in the original paper document**
> Another remark on the well-known endpoint: it should also have the jwks-endpoint, so clients can retrieve the keys needed to validate the ID-tokens that they receive.
> ​​Yet another remark, also the /authorization and /token endpoints should be included. Probably this guide needs to clarify how the values can be set correctly when deploying the app.
I am assuming the URLs depend on deployment details.

Are these concerns covered?

Reviewer: Yes @MarkvanMents These are covered.

Only thing we need to mention in the documentation is regarding well-known endpoint.
We need to provide an example of well-known endpoint
<[app-url>/oidc/.well-known/openid-configuration

Reviewer 2: yes I agree with Sharishma, the well-known endpoint is the only endpoint that needs to be well-documented; it's the only endpoint that is needed at the OIDC-SSO module to connect with the OIDC provider, see https://docs.mendix.com/appstore/modules/oidc/#511-general-oidc-providers

    {
    "post_logout_redirect_uris": [
    "http://localhost:8081/logout"
    ],
    "grant_types": [
    "authorization_code",
    "password"
    ],
    "client_secret_expires_at": 0,
    "scope": "openid",
    "client_secret": "ClientSecret",
    "redirect_uris": [
    "http://localhost:8081/oauth/v2/callback"
    ],
    "client_id_issued_at": 1675940602,
    "backchannel_logout_uri": "http://localhost:8081/logout",
    "client_name": "ClientName",
    "client_id": "ClientID"
    }
    ```

#### 3.3.2 Manual Client Registration
    
If you cannot use automatic registration, you can register the client manually.

1. Select **Manual Registration**.
1. Add the following information:
    * **Client Name**
    * **Client ID**
    * **Alias**
    * **Client Secret** 

>     * **Post Logout redirect URI** – for example, for testing a local OIDC Provider app on port `8081`, `http://localhost:8081/logout` 

Reviewer: Shall we change the urls to <appurl>/logout?

If they try testing OIDC SSO and OIDC Provider locally, then they need to change the ports

Reviewer: POst Logout redirect URI and Redirect URI urls are of OIDC Client port

Author: Sorry, I don't understand this comment.

I thought our example had the OIDC Provider running on 8080 and the OIDC Client running on 8081?

Are you saying that just "/logout" will work and the URIs don't need to be fully qualified? But if "/logout" works in the cloud, then it will also work when testing locally, won't it?

Reviewer: Sorry Mark. a word is missed previous comment
`Shall we change the urls to /logout?`
Shall we change the urls to `<appurl>/logout`   
appurl - OIDC Client url

Author: You mean something like:
```suggestion
    * **Post Logout redirect URI** – the fully qualified logout url, `<appurl>/logout` — for example, for testing a local OIDC Provider app on port `8081`, `http://localhost:8081/logout` 
```

Reviewer: yes @MarkvanMents. I mean the same

But one change here:
Post Logout redirect URI and Redirect URI are the URLs of OIDC SSO.
Can we modify it(for example, for testing a local OIDC Provider app on port `8081`, `http://localhost:8081/logout) as for example, for testing a local OIDC Provider app with OIDC SSO which is on port `8081`, `http://localhost:8081/logout
Same applies to Redirect URI

Here in the examples 
OIDC SSO running on port 8081
OIDC Provider running on port 8080

Author: @SharishmaPodduturi999 - like this? - I've also suggested changing the next line too.
```suggestion
    * **Post Logout redirect URI** – the fully qualified logout url, `<appurl>/logout` — for example, for testing a local OIDC SSO app on port `8081`, `http://localhost:8081/logout`
```

    * **Redirect URI** – for example, for testing a local OIDC Provider app on port `8081`, `http://localhost:8081/oauth/v2/callback`
    * **Back channel logout session support**
    * **Front channel Logout URI**

>     * **Requires PKCE** – this needs to be enabled in both the client and the server to use PKCE

Reviewer 2: On PKCE:
client (based on OIDC SSO) detects if IdP suppoprts PKCE through well-known endpoint.
So in well-known endpoint we should simply state that the OIDC Provider supports it by including the code_challenge_methods_supported on the well-known endpoint. No need to make that configurable - let's always put it there with value "S256". @SharishmaPodduturi999 

This hides the whole PKCE topic fopr the lowcode developer. Highest level of security out-of-the-box; no configurations needed, neither cleint-side or provider-side.

Reviewer: Yes @JaapF ,
Even though it is configurable item on the UI, 
S256 is always included in the well-known endpoint even if it is not enabled.
"code_challenge_methods_supported": [
    "plain",
    "S256"
  ],
I will remove it from UI in this release. So we can remove it in document as well @MarkvanMents

Reviewer 2: 👍

Author: So this can be removed @SharishmaPodduturi999 ?
```suggestion
```

Reviewer: Yes @MarkvanMents ,
this can be removed now

## 4 Configuring an OIDC Client

You need to configure the OIDC SSO module in your app which is using the IAM broker. Perform the following steps to configure the OIDC SSO module in your client app, and test that it is correctly communicating with the IAM broker.

1. Create an app containing the OIDC SSO module as described in [OIDC SSO](/appstore/modules/oidc/).

    {{% alert color="info" %}}

> If you are testing locally, you will need to run your OIDC client on a different port from the IAM broker.

Author: ```suggestion
If you are testing locally, you will need to run your OIDC client in a separate copy of Studio Pro and on a different port from the IAM broker.
```

Clarify what we are doing in local testing?

Reviewer: In local,
We run OIDC Provider on one port ex:8080
and OIDC Client on other port ex:8081
Run the applications parallelly.
1. Open OIDC Provider localhost url, configure client details like clientID, ClientSecret
2. Move to OIDC Client and configure provider details, and try login from OIDC Client

Reviewer 2: I am inclined to say that section "Configuring an OIDC Client" is obsolete. It doubles what is already described in the documentation for OIDC SSO.
@SharishmaPodduturi999 , do you agree? The only thing that should matter is the well-known endpoint which could be local or elsewhere, right?

Author: Done https://github.com/mendix/docs/pull/5616/commits/6b4c1384c138cd39b637eda5db44a88ff7dd2d99

In the [Server](/refguide/configuration/#server) tab of your active configuration, change the **Runtime port** and **Admin port** to be different from those of your IAM broker app. For example, if your IAM broker is running using `8080` and `8090`, then use `8081` and `8091` respectively.
    {{% /alert %}}
    
    1. When you get to [Configuration of OIDC Provider](/appstore/modules/oidc/#oidc-configuration), you already have the values from the previous section.

    1. In the [OIDC Client Configuration](/appstore/modules/oidc/#oidc-configuration) section, add a client configuration and use the following values:

        * **Client ID** – the **Client ID** of the IAM Broker
        * **Client Authentication Method** – *Client ID and Secret*
        * **Client Secret** – the **Client Secret** of the IAM Broker
    

>     1. Enter the **Automatic Configuration URL** using the URL of your IAM Broker app with the well-known (`/oidc/configuration`) endpoint and click **Import Configuration**. For example, when running locally, `http://localhost:8080/oidc/configuration`

Reviewer: @MarkvanMents The well-known configuration endpoint is modified in this current release.
Can you please update it to <[app-url>/oidc/.well-known/openid-configuration

Author: Done https://github.com/mendix/docs/pull/5616/commits/6b4c1384c138cd39b637eda5db44a88ff7dd2d99

1. Sign back in to the app using the OIDC SSO client alias you have just configured.
1. Login by entering credentials of the user which you have created earlier on OIDC provider Accounts section.
    You should be able to login successfully and get into the index.html page 

## 5 Troubleshooting

### 5.1 Infinite Loop of Redirects

The OIDC provider module sets a cookie as a means to persist the session in the user’s browser. If the cookie is not properly set, this may lead to problems. For example, when the OIDC Provider module is used to build an IAM Broker, no session is established and the broker may initiate a new session at the upstream IdP, which results in an ‘infinite loop’ of redirects via the user’s browser.
To ensure the cookie is properly set, the runtime setting com.mendix.core.SameSiteCookies must have value None. See [Environment Details](https://docs.mendix.com/developerportal/deploy/environments-details/#4222-applying-a-different-samesite-setting) for more information how to set the correct value for SameSite runtime setting. Note that the default value for this setting has changed, see [Release Notes](https://docs.mendix.com/releasenotes/studio-pro/8.11/).

> ### 5.2 Custom Claims Are Not Returned Properly

Author: Is this still an issue? There was a discussion in the paper document about whether it should be removed entirely - but it isn't clear that this has been completely resolved.

Reviewer: No Mark, it is not resolved

* It supports the OIDC ‘nonce’ parameter, PKCE, and multiple client authentication methods ( client_secret_post, client_secret_basic) as security features.
* It supports front channel and back-channel logout as public beta features.
    
**Limitations**

* The hybrid and client credential grants are not supported, although the OIDC Provider may contain some (rudimentary) implementation to support them.
* The current implementation is an early [beta release](/releasenotes/beta-features/).

### 1.3 Dependencies

This module requires your app to be using Mendix 9.12.5+ or higher
The following modules need to be imported into your app

* [Community Commons](https://marketplace.mendix.com/link/component/170) – see [Community Commons](/appstore/modules/community-commons-function-library/) documentation
* [Mx Model reflection](https://marketplace.mendix.com/link/component/69) – see [Mx Model Reflection](/appstore/modules/model-reflection/) documentation
* [Administration](https://marketplace.mendix.com/link/component/23513) – see [Administration](/appstore/modules/administration/) documentation

> 

Reviewer: @MarkvanMents 
Can you add one more dependency
[Copy To Clipboard](https://marketplace.mendix.com/link/component/119337)

Author: Done.
Are you aware that this is a **Community Supported** module?
I'm not sure we should be making a **Platform** supported module which has a community supported dependency. @JaapF, what do you think?

I have added it anyway.

Reviewer: @MarkvanMents Yes, I observed that but to complete a task I have used the module. I will try to remove this if possible in the upcoming release.

Reviewer: Hi @MarkvanMents , I have implemented the logic in OIDC Provider itself without depending on any other module. So now we can remove Copy to clipboard dependency.

Author: Removed

## 2 Installation

To install the OIDC Provider module you need to import the following into your app:

> * [OIDC Provider](https://example.com)

Reviewer: @MarkvanMents Add a note saying 
Make sure to remove 'MendixUserID' account entity if exists when using Marketplace version of OIDC Provider rather than a private release of OIDC Provider,

Author: We are documenting the Marketplace version here - surely they just need to import it?

Reviewer: Yes, they need to import it

Author: Not sure what the resolution is here?
@JaapF, I think that if there are beta customers out there, they need a separate bit of documentation rather than having these instructions in the public documentation?

Reviewer 2: Good one Mark, we only have a few of them, so yes let's include that in the release notes for the first version in the marketplace. @SharishmaPodduturi999 can you include this in a (draft) release note?

Author: We agreed in the meeting on 17 April that this information will be provided to beta customers separately.

* [Community Commons](https://marketplace.mendix.com/link/component/170)
* [Mx Model reflection](https://marketplace.mendix.com/link/component/69)
* [Administration](https://marketplace.mendix.com/link/component/23513)

The module then needs to be configured as described below.

## 3 Configuration

### 3.1 Configure Security

To configure the app security, do the following:

1. Open the [User Roles](/refguide/user-roles/) in the app and **Edit** the **Administrator** user role.
1. **Edit** the **Module roles**
1. Assign the following module roles to the Administrator:
    1.  MxModelReflection
        * ModelAdministrator
        * Readonly
        * TokenUser
    1. OpenIDConnectProvider
        * Admin

### 3.2 Configure App

You now need to configure the app.

1. Open the **Domain model** of the **Administration** module.

> 1. Edit the **Account** entity.

Reviewer: @MarkvanMents As per the new task, we need to remove this point and just add

Edit the **Account** entity and give read access to **AccountDetail_Account** association in access rules tab
![image](https://user-images.githubusercontent.com/110099605/217791754-cac80a1e-6860-410c-bb6d-9d07a301e57e.png)
Note: Make sure to do this if you upgrade to the latest version of OIDC Provider

Author: So the customer doesn't have to add the new attribute? There is already an association there?

Author: See changes below 👇

Reviewer: > So the customer doesn't have to add the new attribute? There is already an association there?

Yes they don't need to add the attribute 'mendixUserID'

    1. Add a new attribute as follows:
        * **Name** – MendixUserID
        * **Type** – **AutoNumber**
    1. Switch to the **Access Rules** tab.
    1. **Edit** the **Administrator** module role.
    1. Give **Read** **Access rights** to the **MendixUserID** attribute.
1. Open the **Runtime** tab of the app **Settings**.
1. Add the **OpenIDConnectProvider.ASU_Start_All** microflow as an [After Startup](/refguide/app-settings/#after-startup) microflow.
1. Finally, ensure that your app has the following Administration navigation items:

    | Caption | Action | User Roles |
    | --- | --- | --- |

>     | Accounts | Open page 'Administration.Account_Overview' | Administrator, UserManager |

Reviewer: Can you please add a point here saying 
If you are using Marketplace version of OIDC Provider rather than a private release of OIDC Provider, then 
Accounts|Open Page 'OpenIDConnectProvider.AccountDetail_Overview'|Administrator

Reviewer: Make sure to create accounts using OpenIDConnectProvider.AccountDetail_Overview rather than Administration.Account_Overview page

Author: Changed, I am documenting the Marketplace version,

    | OpenID Connect | Open page 'OpenIDConnectProvider.OpenIDConnect_Dashboard' | Administrator |

### 3.3 Configure OIDC Provider (This App)

The rest of the configuration can be performed through the app.

1. Deploy and run the app.
1. Login as an administrator.
1. Use the **Accounts** navigation item.
1. Click **New local user** and create a new active account with the **User role** set to **User**.
1. Use the **OpenID Connect** navigation item.
1. Open the **Server Keys** tab.
1. Click **New Key**.
1. Open the **Scopes** tab.
1. Create a **New** scope with the **Scope type** set to **Standard**.
1. Open the **Clients** tab.
1. Click **New Client Registration**.

    You can register a new client in one of two ways. Follow one of the sets of instructions below:

#### 3.3.1 Automatic Registration

To check that this works, you will need a tool for testing APIs (such as [Postman](https://www.postman.com/)) which allows you to execute HTTP commands to the Registration URI of you IdP.

1. Select **Automatic Registration**.
1. Store the **Registration Access token** and the **Registration URI**.
1. In your API testing tool, set up the following:

>     * a **POST** command to the **Registration URI**

Reviewer: Can we add example registration URI
http://localhost:8080/oidc/register

>     * an HTTP header with **Key** = "AUthorization" and **Value** = **Registration Access token**

Reviewer: an HTTP header with **Key** = "AUthorization" and **Value** = **Bearer Registration Access token**

1. Send the command. If it is working, you should get a response which resembles the example below:

> 

Reviewer: the response will be as shown below
{
    "post_logout_redirect_uris": [
        "http://localhost:8081/logout"
    ],
    "grant_types": [
        "authorization_code",
        "password"
    ],
    "client_secret_expires_at": 0,
    "scope": "openid",
    "client_secret": "c46591bd-8fae-4f90-9efb-b5973bea04df",
    "redirect_uris": [
        "http://localhost:8081/oauth/v2/callback"
    ],
    "client_id_issued_at": 1675940602,
    "backchannel_logout_uri": "http://localhost:8081/logout",
    "client_name": "DemoClient1",
    "client_id": "DemoClient1"
}

>     ```json {linenos=false}

Reviewer: This json will be the body of the request @MarkvanMents

Author: This is the response, not the request?

Reviewer: Yes Mark. The json commented is response

Reviewer: This should be the body of request

{
  "client_id" : "DemoClient2",
  "client_name" : "DemoClient2",
  "client_secret" : "c46591bd-8fae-4f90-9efb-b5973bea04df",
  "redirect_uris" : [ http://localhost:8081/oauth/v2/callback ],
  "backchannel_logout_uri"    : http://localhost:8081/logout,
  "post_logout_redirect_uris" : [http://localhost:8081/logout],
  "grant_types": [  "authorization_code","password" ],
  "scope":  "openid" 
}

Author: Done

    {
        "client_id" : "DemoClient1",
        "client_name" : "DemoClient1",
        "client_secret" : "DemoCLient1",
        "redirect_uris" : [ http://localhost:8081/oath/v2/callback ],
        "backchannel_logout_uri" : http://localhost:8081/logout,
        "post_logout_redirect_uris" : [ http://localhost:8081/logout],

>         "grant_types" : [ "hybrid", "client_credentials", "password" ],

Reviewer: In the grant types we decided to include "authorization_code","password" only
"grant_types" : [ "authorization_code","password" ],

        During development and testing, you can use the local host port number according to your case. Example: If your OIDC client app port is 8080 then use http://localhost:8080/oauth/v2/callback and if your OIDC client port is 8081 then use [http://localhost:8081/oauth/v2/callback](http://localhost:8081/oauth/v2/callback). It is 8081 in my case. 

### 3.4 Configure OIDC Client

This section describes configuration steps a low-code developer needs to do to run and test the interaction between his ‘IdP-app’ and his ‘client app’ using OIDC SSO module.

* Open OIDC Client in Mendix Studio pro. Change the local host port number as OpenIDConnectProvider runs on 8080. 
![](https://paper-attachments.dropbox.com/s_A27DA9F06997C3D919A20D2E1F62DBDD4D4A86848C0F5AA1A9B4E4528436EC3B_1660494287240_file.png)

* Run the app locally. Login as demo administrator. Add the OIDC provider configurations on OIDC Client end 
    ![](https://paper-attachments.dropbox.com/s_A27DA9F06997C3D919A20D2E1F62DBDD4D4A86848C0F5AA1A9B4E4528436EC3B_1660494287247_file.png)

    * Client secret- c46591bd-8fae-4f90-9efb-b5973bea04df 

    * Client ID- DemoClient 

>     * Welknown-openid configuration URL: [http://localhost:8080/oidc/configuration](http://localhost:8080/oidc/configuration) 

Reviewer: I am about to test project by changing well-known configuration endpoint. I will add that URL once I complete testing

    * Add scopes: email, profile, OpenID 

        ![](https://paper-attachments.dropbox.com/s_A27DA9F06997C3D919A20D2E1F62DBDD4D4A86848C0F5AA1A9B4E4528436EC3B_1660494287255_file.png)

    ![](https://paper-attachments.dropbox.com/s_A27DA9F06997C3D919A20D2E1F62DBDD4D4A86848C0F5AA1A9B4E4528436EC3B_1660494287263_file.png)

* Now sign-out from OIDC Client and login with DemoClient. 
    ![](https://paper-attachments.dropbox.com/s_A27DA9F06997C3D919A20D2E1F62DBDD4D4A86848C0F5AA1A9B4E4528436EC3B_1660494287268_file.png)

* Login by entering credentials of the user which you have created earlier on OIDC provider Accounts section 
    ![](https://paper-attachments.dropbox.com/s_A27DA9F06997C3D919A20D2E1F62DBDD4D4A86848C0F5AA1A9B4E4528436EC3B_1660494287274_file.png)

* You should be able to login successfully and get into the index.html page 
    ![](https://paper-attachments.dropbox.com/s_A27DA9F06997C3D919A20D2E1F62DBDD4D4A86848C0F5AA1A9B4E4528436EC3B_1660494287278_file.png)

> ## 4 Testing OIDC Provider with Postman 

Reviewer: @MarkvanMents we can remove this section as this is not low code testing

{{< figure src="/attachments/appstore/modules/oidc-provider/typical-usage.png" >}}

You can also set up your users manually in your app, using the Mendix Administration module, rather than linking to your central IdP.

### 1.1 Typical Usage Scenarios

The following are usage scenarios that would be achievable with the OIDC Broker.

* Mendix customers that want to build an IAM Broker solution that would hide the complexity of a multitude of Mendix apps from their corporate IdP.  By having those apps delegate authentication to the broker and have the broker delegate authentication to their IdP, only one OAuth client needs to be configured at their IdP.  A deployment pipeline (deployment agent) can register additional Mendix apps with the IAM Broker in an automated fashion via an API. 
* Mendix Solution Vendors (MSVs) who want to simplify the microservice architecture of their solution from the customer by using the IAM Broker as a single IAM integration point for their customers.

### 1.2 Features and Limitations

The OIDC Provider has the following features and limitations:

**Features**

* It supports responsive web applications, using the common OAuth Authorization Code grant.

> * Your apps can be registered as an OIDC client with the OIDC Provider using the client registration API or client configuration screen.

Reviewer 2: Let's clarify one step further:
"Usage of client registration API is useful when you want your CICD deployment pipeline to register your client in an automated fashion."

Author: @JaapF - does this look OK?
```suggestion
* Your apps can be registered as an OIDC client with the OIDC Provider using the client registration API or client configuration screen. The client registration API allows you to register your client automatically when using a CI/CD deployment pipeline.
```

Reviewer 2: 👍

* It publishes a well-known endpoint to communicate endpoints and other IdP characteristics to client applications. This can be used by clients to retrieve the keys needed to validate ID-tokens that they receive.
* It supports the OIDC ‘nonce’ parameter, PKCE, and multiple client authentication methods ( client_secret_post, client_secret_basic) as security features.
    
**Limitations**

> * The hybrid and client credential grants are not supported, although the OIDC Provider may contain some (rudimentary) implementation to support them.

Reviewer 2: Some more limitations I'd like to add.
- The OIDC Provider module ignores "email", "phone" and "profile" scope values (as specified by OIDC specs) when the client includes these in a auth*n-request . Instead, the OIDC Provider module will include user claims in an ID-token based on a custom microflow, regardless of the scopes in the request.

Author: @JaapF - does this look OK. Also includes information about support status of logout.
```suggestion
* The hybrid resource owner password credentials, and client credential grants are not supported, although the OIDC Provider may contain some (rudimentary) implementation to support them.
* The OIDC Provider module ignores "email", "phone" and "profile" scope values (as specified by OIDC specs) when the client includes these in a authentication request . Instead, the OIDC Provider module will include user claims in an ID-token based on a custom microflow, regardless of the scopes in the request.
* Front channel and back-channel logout are implemented as alpha features.
```

Reviewer 2: @MarkvanMents can we indicate the front&back channel logout as 'alpha' implementation rather than beta.
It is not yet tested by the developer and it is probably not feature complete

Reviewer 2: @MarkvanMents can we say:
"The hybrid, resource owner password credentials and client credential grants are not supported, ..."

1. Finally, ensure that your app has the following Administration navigation items:

    | Caption | Action | User Roles |
    | --- | --- | --- |
    | Accounts | Open page 'OpenIDConnectProvider.AccountDetail_Overview' | Administrator |
    | Active Sessions | Open page 'Administration.ActiveSessions' | Administrator |
    | Scheduled Events | Open page 'Administration.ScheduledEvents' | Administrator |
    | Runtime Instances | Open page 'Administration.RunteimInstances' | Administrator |
    | My Account | Call microflow 'Administration.ManageMyAccount' | User |
    | OpenID Connect | Open page 'OpenIDConnectProvider.OpenIDConnect_Dashboard' | Administrator |

### 3.3 Configure OIDC Provider

The rest of the configuration can be performed through the app.

1. Deploy and run the app.
1. Login as an administrator.
1. Use the **Accounts** navigation item.
1. Click **New** and create the user account(s) you need in your IAM Broker — for example, a new active account with the **User role** set to **User**.
1. Use the **OpenID Connect** navigation item.
1. Open the **Server Keys** tab.
1. Click **New Key**.
1. Open the **Scopes** tab.

> 1. Create a **New** scope with the **Scope type** set to **Standard**.

Author: Should this point to the new sections below about configuring centralized authorization with Scopes? Then the customer will have a lot more information about what they should be doing with scopes.

Reviewer 2: yes it should!

Author: @JaapF - does this look OK?
```suggestion
1. Open the **Scopes** tab.
1. Create a **New** scope with the **Scope type** set to **Standard**. See [Configuring Centralized Authorization](#configuring-authorization) for more information on what scopes you should set.
```

    "client_secret": "ClientSecret",
    "redirect_uris": [
    "http://localhost:8081/oauth/v2/callback"
    ],
    "client_id_issued_at": 1675940602,
    "backchannel_logout_uri": "http://localhost:8081/logout",
    "client_name": "ClientName",
    "client_id": "ClientID"
    }
    ```

#### 3.3.2 Manual Client Registration
    
If you cannot use automatic registration, you can register the client manually.

1. Select **Manual Registration**.
1. Add the following information:
    * **Client Name** – a name for this client so that it is easy to identify
    * **Client ID** – a unique string which identifies this client
    * **Alias**
    * **Client Secret** – a secure string to allow the client to authenticate to the OIDC Provider module
    * **Post Logout redirect URI** – for example, for testing a local OIDC Provider app on port `8081`, `http://localhost:8081/logout` 
    * **Redirect URI** – for example, for testing a local OIDC Provider app on port `8081`, `http://localhost:8081/oauth/v2/callback`
    * **Back channel logout session support**
    * **Front channel Logout URI**
    * **Requires PKCE** – this needs to be enabled in both the client and the server to use PKCE

> 

Reviewer 2: @Mark, based on my conversation with Sharishma earlier today I think we need to include a section on user provisioning; to be precisely: on user objects.

Sharishma wrote a draft for that, which oi have slightly enhanced, see this new section in the dropboxpaper we have used before: https://paper.dropbox.com/doc/OIDC-Provider-scopescustom-claims--B2xPiByXGfTye6IdUKEfjfUuAg-YSTHWbqkRXiVB4gPY2MyH#:uid=277027133043496771341533&h2=Account-creation-in-OIDC-Provi

Author: Added this in the following commits: https://github.com/mendix/docs/pull/5616/files/ab80941f505533587620f6a41c3bf3ba4bf3aab4..589719a31b77e3d50b55416aa9d52cffd51db1ec

Choose one of the two options, below. 

Whichever option you choose, you will need to use [Custom User Provisioning](/appstore/modules/oidc/#custom-provisioning) in the OIDC SSO module of your client app to assign the correct user roles to the end-user.

##### 3.3.3.1 Configuration of the OIDC Provider for Centralized Authorization with Scopes

 When you follow the recommended approach for [centralized authorization](#centralized-auth) using OAuth scopes, the OIDC Provider needs to know which user roles have been implemented in your Mendix apps, so the client can assign those user roles.

The scope values (user roles) that you assign to the authenticated user in the OIDC Provider are sent to your app in both the ID-token and Access Token. Communicating authorization via Access Token is the standard approach when using OAuth. If your client app consumes APIs, it can forward the Access Token as a bearer token in the http authorization header in such API calls.
 
To return requested scopes to your client app, you need to perform the following steps to configure the OIDC Provider:

> 1. Open the **Scopes** tab for the client you want to configure

Author: Is the scope defined per client or for the whole app?
In section 3.3 it says to set up scopes _before_ registering the client.

Reviewer 2: @SharishmaPodduturi999 can you answer this one.

My understanding is:
1 - we configure a list of scopes for all apps together
2 - as a second step we associate with clients

I'm not sure, please confirm - maybe sharing a screenshot of the UI-structure in the dropbox paper helps Mark and myself understand.

Reviewer: Yes @JaapF You are correct
1. We configure list of scopes for all apps together
I am adding screenshot here
![image](https://user-images.githubusercontent.com/110099605/233281765-3764d1b5-ba7c-4e43-9e93-9402974d599f.png)

2.Associating scope to specific clients
![image](https://user-images.githubusercontent.com/110099605/233282101-d2a9c013-45c7-44b5-98eb-278ad2513d69.png)

Author: Added to section 3.3 in this commit: https://github.com/mendix/docs/pull/5616/commits/9487ff5d4aa25a47be16995f4a677c81ec8089d5

1. Create a new scope for every user role which is implemented in your client apps. This scope has the same **Name** as the user role in your client app.

    If your client app consumes APIs and wants to get access using the Access Token from the OIDC Provider, you may want to configure additional scope values, as required by those APIs.

> 

Reviewer 2: FYI. This is a reason why the scopes are configured as a single list for all apps rtaher than 'per app'. In a microservice landscape scopes may be shared between multiple apps.
No need to clarify this in the doc.

    
    {{% alert color="info" %}}The default implementation of `OpenIDConnectProvider.SUB_CustomScope` grants any requested scope provided it is associated with client (your Mendix app). It does not do any validation of the end user.{{% /alert %}}

##### 3.3.3.2 Configuration of the OIDC Provider to Propagate the End-User’s Identity with Custom Claims

Typically you want to propagate the end-user’s identity from the OIDC Provider to your Mendix app. Although the basic user attributes like ‘email address’ and ‘user name’ may be sufficient, your app may need more information about the end-user. User attributes like ‘department’ or ‘job-title’ may be used for business logic, including decentralized authorization.

To pass this additional information, you need to create custom claims. You can do this as follows:

1. Create a microflow which returns a value to the claim.
1. Run (publish) your app.
1. Sign in to your app as an Administrator.

    {{% todo %}}Where was MxObjects set up? I don't think we have any navigation to this page set up in 3.2?{{% /todo %}}

> 1. Open the MxObjects overview page and synchronize the required modules to see the new microflow.

Author: We never set this up in the navigation. What is the MxObjects overview page and how does the customer get to it?

Reviewer 2: Based on the meeting we had earlier this week (or was it last week?) my understanding was that this has to do with ModelReflection module. @SharishmaPodduturi999 , plesae clarify

Reviewer: @JaapF It has to do with MxModelReflection module

@MarkvanMents 
Sorry missed this point
"MxModelReflection.MxObjects_Overview" page needs to be added in the navigation to syncronize  the required modules to get created new microflows

Author: Added in commit https://github.com/mendix/docs/pull/5616/commits/9487ff5d4aa25a47be16995f4a677c81ec8089d5

1. Follow the navigation item OpenID Connect to open the page `OpenIDConnectDashboard`.
1. Switch to the Custom claims tab of your registered client.
1. Create a new claim.
    * Provide a name for claim
    * Select the microflow which returns the value to the claim
1. Save the claim.

    When an authentication request is made to the OIDC Provider, the created custom claim will be added to the ID-token.

{{% alert color="info" %}}
Do not to create custom claims with the same names as standard claims or existing Mendix custom claims.

Some examples of existing claims are:
* scope

    If the claim with ‘scope’ is created, the scope value in access token and id-token gets overwritten with the claim value

> * …..aud

Author: This looks strange - is it correct?

Reviewer 2: @MarkvanMents you mean the dots?
"aud" is one of the claims that are included in Access Tokens and/or ID tokens.

Author: Yes, I mean the dots. Are they part of the scope? It is particularly strange to have an ellipsis (…) as part of the scope, as it isn't an ASCII character.

Reviewer 2: the name of the claim is simply aud, so please remove the dots. Sorry for not being clear enough in my previous comment.

Author: ```suggestion
* aud
```

* sub
* iss
* name
* exp
* iat
* nonce
* com.mendix.user.language
* com.mendix.user.entity
* com.mendix.user.attributes
* com.mendix.user.timezone

> * com.mendix.user.roles

Reviewer 2: @SharishmaPodduturi999 does the OIDC Provider module really generate these claims?
How does it get values for these claims?

Please advise what is implemented. I would prefer to follow standards rather than use proprietary claims (the standardized equivalent cliam names i would have to look-up).
Does the OIDC SSO module uses these claims when they are included?

Reviewer: @JaapF 
`does the OIDC Provider module really generate these claims?` - yes OIDC Provider generate the claims

Access token 
"aud": "DemoClient" - clientID
 "sub": "3" - An autonumber(identifier of user who tries to login)
 "nbf": 1681982966,
 "scope": "",
 "iss": "http://localhost:8080/",
 "name": "test" - username of the user
 "exp": 1682069366,
 "iat": 1681982966

ID-token
"com.mendix.user.language": "en_US" - language selected for the user who tries to login
  "sub": "test" - username of the user who tries to login
  "iss": "http://localhost:8080/" - issuer/OIDc Provider app URL
  "com.mendix.user.entity": "Administration.Account" : entity of the user who tries to login
  "nonce": "zjws1eFqMuxA2sYQrT6F" : nonce value from request
  "com.mendix.user.attributes": { - adding user attributes which have read access and it does not include virtual attributes and password
    "Email": "",
    "FullName": "test",
    "IsLocalUser": "true"
  },
  "aud": "DemoClient" - clientID
  "scope": "" - scopes(userroles)
  "exp": 1681983561,
  "com.mendix.user.timezone": "" -includes users timezone if set
  "iat": 1681982961,
  "com.mendix.user.roles": [ - roles of the user who tries to login
    "User"
  ]
  
  
  
  **Does the OIDC SSO module uses these claims when they are included?**
Only name,email and sometimes sub are used in OIDC SSO

Reviewer 2: @MarkvanMents , cc @SharishmaPodduturi999 
Let's add the following info to the document.

"The ID-token contains a few claims that have a name that starts with com.mendix.user. These claims are an alpha feature and may be deprecated in a future release in favour of standard claims specified by OIDC protocol. The OIDC SSO module does not process these claims."

Author: Added in commit: https://github.com/mendix/docs/pull/5616/commits/9487ff5d4aa25a47be16995f4a677c81ec8089d5

1. Sign back in to the app using the OIDC SSO client alias you have just configured.
1. Login by entering credentials of the user which you have created earlier on OIDC provider Accounts section.
    You should be able to login successfully and get into the index.html page 

## 5 Troubleshooting

### 5.1 Infinite Loop of Redirects

The OIDC provider module sets a cookie as a means to persist the session in the user’s browser. If the cookie is not properly set, this may lead to problems. For example, when the OIDC Provider module is used to build an IAM Broker, no session is established and the broker may initiate a new session at the upstream IdP, which results in an ‘infinite loop’ of redirects via the user’s browser.
To ensure the cookie is properly set, the runtime setting com.mendix.core.SameSiteCookies must have value None. See [Environment Details](https://docs.mendix.com/developerportal/deploy/environments-details/#4222-applying-a-different-samesite-setting) for more information how to set the correct value for SameSite runtime setting. Note that the default value for this setting has changed, see [Release Notes](https://docs.mendix.com/releasenotes/studio-pro/8.11/).

> ### 5.2 Custom Claims Are Not Returned Properly

Reviewer: @MarkvanMents This is fixed So we can remove this

This is a known issue and will be corrected in a future release.

## 6 About Authorization

When you are building Mendix apps you need to make two architectural decisions in the area of end-user access:

* Which component in the application landscape will authenticate end-users
* Which component in the application landscape will authorize end-users to use a the user roles of your app

When you use the OIDC Provider module, you have already taken the decision that your app will delegate user authentication to the app you build with the OIDC Provider module.
With respect to authorization that leaves you, again, with two options:

* Fully decentralized authorization
* Centralized authorization

These two options are described below.

> ### 6.1 Fully Decentralized Authorization

Author: Should we have Centralized Authentication as the first option, since that includes the recommended use and the one we describe in detail above?

Reviewer 2: Yes, agreed.

The OIDC Provider module allows you to pass user attributes in the ID-tokens that are sent from the OIDC Provider to your app. You can define ‘custom’ claims, so the user attributes that can be communicated are not restricted to a pre-defined set.

### 6.2 Centralized Authorization {#centralized-auth}

With centralized authorization your app delegates authorization decisions to a central component which then communicates that decision to your app. Your app then enforces the authorization decision by applying one or more user roles to the end-user, but does not need any business logic to decide which user roles to apply.

The OIDC Provider module is one such central component and you can communicate the authorization decision made by the OIDC Provider module in 2 ways:

* using the concept of OAuth scopes (recommended)
* using a custom user claim

> Using OAuth scopes is the recommended approach since it is the standard OAuth solution. With Mendix, we advise you to think of an app’s user roles as being the same as OAuth scope values.  By adhering to this logic, you can develop apps with any user roles without having to decide and agree on custom attributes. You can customize the OIDC SSO module with microflows which parse the tokens from the OIDC Provider module and apply user roles to enforce the authorization indicated in the token.

Author: I've taken out references to out-of-the-box token parsing microflows since the customer has to write a custom one.

Reviewer 2: ok agreed.
Once we have those available, we'll do a small PR.
https://paper.dropbox.com/doc/How-to-assign-userroles-based-on-Access-Token-andor-ID-tokens--B2tpNWGRFcBEUC74hAjUh3hsAg-rQgdmYqREcdhloAgBCR8D

    ```json {linenos=false}
    {
    "post_logout_redirect_uris": [
    "http://localhost:8081/logout"
    ],
    "grant_types": [
    "authorization_code",
    "password"
    ],
    "client_secret_expires_at": 0,
    "scope": "openid",
    "client_secret": "ClientSecret",
    "redirect_uris": [
    "http://localhost:8081/oauth/v2/callback"
    ],
    "client_id_issued_at": 1675940602,
    "backchannel_logout_uri": "http://localhost:8081/logout",
    "client_name": "ClientName",
    "client_id": "ClientID"
    }
    ```

#### 3.3.2 Manual Client Registration
    
If you cannot use automatic registration, you can register the client manually.

1. Select **Manual Registration**.
1. Add the following information:
    * **Client Name** – a name for this client so that it is easy to identify
    * **Client ID** – a unique string which identifies this client
    * **Alias** – usually the same as *Client Name* but can be different
    * **Client Secret** – the client password to allow the client to authenticate to the OIDC Provider module
    * **Post Logout redirect URI** – for example, for testing a local OIDC Provider app on port `8081`, `http://localhost:8081/logout` 

>     * **Redirect URI** – for example, for testing a local OIDC Provider app on port `8081`, `http://localhost:8081/oauth/v2/callback`

Author: @SharishmaPodduturi999 - this change too?
```suggestion
    * **Redirect URI** – for example, for testing a local OIDC SSO app on port `8081`, `http://localhost:8081/oauth/v2/callback`
```

        "Email": "",
        "FullName": "test",
        "IsLocalUser": "true"
    },
    "aud": "DemoClient",
    "scope": "",
    "exp": 1681970318,
    "com.mendix.user.timezone": "",
    "iat": 1681969718,
    "com.mendix.user.roles": [
        "User"
    ]
}
```

## 6 Troubleshooting

### 6.1 Infinite Loop of Redirects

The OIDC provider module sets a cookie as a means to persist the session in the user’s browser. If the cookie is not properly set, this may lead to problems. For example, when the OIDC Provider module is used to build an IAM Broker, no session is established and the broker may initiate a new session at the upstream IdP, which results in an ‘infinite loop’ of redirects via the user’s browser.
To ensure the cookie is properly set, the runtime setting com.mendix.core.SameSiteCookies must have value None. See [Environment Details](https://docs.mendix.com/developerportal/deploy/environments-details/#4222-applying-a-different-samesite-setting) for more information how to set the correct value for SameSite runtime setting. Note that the default value for this setting has changed, see [Release Notes](https://docs.mendix.com/releasenotes/studio-pro/8.11/).

> ### 6.2 Custom Claims Are Not Returned Properly

Reviewer: @MarkvanMents 
Can you please remove 6.2 section as this issue is fixed

---
title: "OIDC Provider"
url: /appstore/modules/oidc-provider/
category: "Modules"
description: "Describes the configuration and usage of the OIDC Provider module, which is available in the Mendix Marketplace."
tags: ["marketplace", "marketplace component", "OIDC", "IdP", "identity provider", "platform support", "Provider"]
#If moving or renaming this doc file, implement a temporary redirect and let the respective team know they should update the URL in the product. See Mapping to Products for more details.
# Linked from https://marketplace.mendix.com/link/component/+++++++
---

## 1 Introduction

{{% todo %}}Add link to module in Marketplace{{% /todo %}}

The [OIDC Provider](https://example.com) can be used to build a Mendix app that acts as an Identity provider(IdP) for other apps. This app then acts as an IAM broker, providing a Single Sign-On (SSO) experience for the end-users of those applications.

> The module supports responsive browser-based applications and has been tested with applications that use the OIDC SSO module. This module can be used in Mendix version 9.12.5 and above.

Reviewer: @MarkvanMents 
Can you please change the version here to 9.22.0?

Author: ```suggestion
The module supports responsive browser-based applications and has been tested with applications that use the OIDC SSO module. This module can be used in Mendix version 9.22.0 and above.
```

The OIDC Provider has the following features and limitations:

**Features**

* It supports responsive web applications, using the common OAuth Authorization Code grant.
* Your apps can be registered as an OIDC client with the OIDC Provider using the client registration API or client configuration screen.
* It publishes a well-known endpoint to communicate endpoints and other IdP characteristics to client applications. This can be used by clients to retrieve the keys needed to validate ID-tokens that they receive.
* It supports the OIDC ‘nonce’ parameter, PKCE, and multiple client authentication methods ( client_secret_post, client_secret_basic) as security features.
    
**Limitations**

* The hybrid and client credential grants are not supported, although the OIDC Provider may contain some (rudimentary) implementation to support them.
* The current implementation of setting custom claims in the generated ID-tokens is not working properly in this release as the custom attribute names that you would define is incorrectly set in the ID-token. If you want to set custom claims, you will have to customize the OIDC Provider module.

### 1.3 Dependencies

> This module requires your app to be using Mendix 9.12.5+ or higher

Reviewer: @MarkvanMents 
Can you please change version 9.12.5+ to 9.22.0

Author: ```suggestion
This module requires your app to be using Mendix 9.22.0 or higher
```

### 1.2 Features and Limitations

The OIDC Provider has the following features and limitations:

**Features**

* It supports responsive web applications, using the common OAuth Authorization Code grant.
* Your apps can be registered as an OIDC client with the OIDC Provider using the client registration API or client configuration screen.
* It publishes a well-known endpoint to communicate endpoints and other IdP characteristics to client applications. This can be used by clients to retrieve the keys needed to validate ID-tokens that they receive.
* It supports the OIDC ‘nonce’ parameter, PKCE, and multiple client authentication methods ( client_secret_post, client_secret_basic) as security features.
    
**Limitations**

* The hybrid and client credential grants are not supported, although the OIDC Provider may contain some (rudimentary) implementation to support them.
* The current implementation of setting custom claims in the generated ID-tokens is not working properly in this release as the custom attribute names that you would define is incorrectly set in the ID-token. If you want to set custom claims, you will have to customize the OIDC Provider module.

> 

Reviewer 2: @MarkvanMents this is no longer a limitation. Custom claims are working.
I think @SharishmaPodduturi999 confirmed that somewhere, I forgot where.

Reviewer: yes @JaapF This is fixed
@MarkvanMents custom claims are working

Author: Removed in suggestion above.

#### 3.3.1 Automatic Client Registration

To check that this works, you will need a tool for testing APIs (such as [Postman](https://www.postman.com/)) which allows you to execute HTTP commands to the Registration URI of you IdP.

1. Select **Automatic Registration**.
1. Take a copy of the **Bearer Registration Access token** and the **Registration URI**.
1. In your API testing tool, set up the following:
    * a **POST** command to the **Registration URI** with the endpoint `/oidc/register`, for example, for a locally-deployed app, `http://localhost:8080/oidc/register`
    * an HTTP header with **Key** = "Authorization" and **Value** = **Bearer Registration Access token**
    * an HTTP request body with the following format (for a Client `ClientID` which is running on host and port `localhost:8081`):

        ```json {linenos=false}
        {
            "client_id" : "ClientID",
            "client_name" : "ClientName",
            "client_secret" : "ClientSecret",
            "redirect_uris" : [ http://localhost:8081/oauth/v2/callback ],
            "backchannel_logout_uri" : http://localhost:8081/logout,
            "post_logout_redirect_uris" : [http://localhost:8081/logout],

>             "grant_types": [ "authorization_code","password" ],

Reviewer 2: @SharishmaPodduturi999 , we don't support the Resource Owner password credential grant, so "Password" needs to be removed from the list of supported grant_types.
(I didn't notice this before)

Reviewer: Yes @JaapF You are correct. Even I missed this point in review.
I have changed it and tested the endpoint now.
new body format is-->
{
  "client_id" : "DemoClient",
  "client_name" : "DemoClient",
  "client_secret" : "c46591bd-8fae-4f90-9efb-b5973bea04df",
  "redirect_uris" : [ http://localhost:8081/oauth/v2/callback ],
  "backchannel_logout_uri"    : http://localhost:8081/logout,
  "post_logout_redirect_uris" : [http://localhost:8081/logout],
  "grant_types": [ "authorization_code" ],
  "scope":  "User" 
}

Author: ```suggestion
            "grant_types": [ "authorization_code" ],
```

            "scope": "openid"
        }
        ```

1. Send the command. If it is working, you should get a response which has a body which resembles the example below:

    ```json {linenos=false}
    {
    "post_logout_redirect_uris": [
    "http://localhost:8081/logout"
    ],
    "grant_types": [
    "authorization_code",

>     "password"

Reviewer 2: same as previous comment.
We need to remove "password" from list of grant_types.

Reviewer: Yes
new response after changing scope and grant types is ->
{
    "post_logout_redirect_uris": [
        "http://localhost:8081/logout"
    ],
    "grant_types": [
        "authorization_code"
    ],
    "client_secret_expires_at": 0,
    "scope": "User",
    "client_secret": "c46591bd-8fae-4f90-9efb-b5973bea04df",
    "redirect_uris": [
        "http://localhost:8081/oauth/v2/callback"
    ],
    "client_id_issued_at": 1682415842,
    "backchannel_logout_uri": "http://localhost:8081/logout",
    "client_name": "DemoClient",
    "client_id": "DemoClient"
}

Author: ```suggestion
```

    | --- | --- | --- |
    | Accounts | Open page 'OpenIDConnectProvider.AccountDetail_Overview' | Administrator |
    | Active Sessions | Open page 'Administration.ActiveSessions' | Administrator |
    | Scheduled Events | Open page 'Administration.ScheduledEvents' | Administrator |
    | Runtime Instances | Open page 'Administration.RunteimInstances' | Administrator |
    | My Account | Call microflow 'Administration.ManageMyAccount' | User |
    | OpenID Connect | Open page 'OpenIDConnectProvider.OpenIDConnect_Dashboard' | Administrator |
    | Mx Objects | Open page 'MxModelReflection.MxObjects_Overview' | Administrator |

### 3.3 Configure OIDC Provider

The rest of the configuration can be performed through the app.

1. Deploy and run the app.
1. Login as an administrator.
1. Use the **Accounts** navigation item.
1. Click **New** and create the user account(s) you need in your IAM Broker — for example, a new active account with the **User role** set to **User**.
1. Use the **OpenID Connect** navigation item.
1. Open the **Server Keys** tab.
1. Click **New Key**.
1. Open the **Scopes** tab.
1. Create **New** scopes with the **Scope type** set to **Standard**.

>     This page sets up a single list of all the scopes that are known to the OIDC Provider module. For each client, you can then choose the scopes used by that specific client.

Author: ```suggestion
    This page sets up a single list of all the scopes that are known to the OIDC Provider module. For each client, you can then choose the scopes used by that specific client. See [Configuring Centralized Authorization](#configuring-authorization) for more information on what scopes you should set.
```

        "Email": "",
        "FullName": "test",
        "IsLocalUser": "true"
    },
    "aud": "DemoClient",
    "scope": "",
    "exp": 1681970318,
    "com.mendix.user.timezone": "",
    "iat": 1681969718,
    "com.mendix.user.roles": [
        "User"
    ]
}
```

## 6 Troubleshooting

### 6.1 Infinite Loop of Redirects

The OIDC provider module sets a cookie as a means to persist the session in the user’s browser. If the cookie is not properly set, this may lead to problems. For example, when the OIDC Provider module is used to build an IAM Broker, no session is established and the broker may initiate a new session at the upstream IdP, which results in an ‘infinite loop’ of redirects via the user’s browser.
To ensure the cookie is properly set, the runtime setting com.mendix.core.SameSiteCookies must have value None. See [Environment Details](https://docs.mendix.com/developerportal/deploy/environments-details/#4222-applying-a-different-samesite-setting) for more information how to set the correct value for SameSite runtime setting. Note that the default value for this setting has changed, see [Release Notes](https://docs.mendix.com/releasenotes/studio-pro/8.11/).

> ### 6.2 Custom Claims Are Not Returned Properly
> 
> This is a known issue and will be corrected in a future release.
> 

Author: ```suggestion
```

* It publishes a well-known endpoint to communicate endpoints and other IdP characteristics to client applications. Mendix apps using the OIDC SSO module will consume this endpoint to perform actions like retrieving the keys needed to validate ID-tokens that they receive.
* It supports the OIDC ‘nonce’ parameter, PKCE, and multiple client authentication methods ( client_secret_post, client_secret_basic) as security features.
    
**Limitations**

* The hybrid resource owner password credentials, and client credential grants are not supported, although the OIDC Provider may contain some (rudimentary) implementation to support them.
* The OIDC Provider module ignores "email", "phone" and "profile" scope values (as specified by OIDC specs) when the client includes these in a authentication request . Instead, the OIDC Provider module will include user claims in an ID-token based on a custom microflow, regardless of the scopes in the request.
* Front channel and back-channel logout are implemented as alpha features.

> * The current implementation of setting custom claims in the generated ID-tokens is not working properly in this release as the custom attribute names that you would define is incorrectly set in the ID-token. If you want to set custom claims, you will have to customize the OIDC Provider module.

Author: ```suggestion
```

            "redirect_uris" : [ http://localhost:8081/oauth/v2/callback ],
            "backchannel_logout_uri" : http://localhost:8081/logout,
            "post_logout_redirect_uris" : [http://localhost:8081/logout],
            "grant_types": [ "authorization_code" ],

>             "scope": "openid"

Reviewer: @MarkvanMents Can you please change the scope as "User" as we are referring scopes as userroles
"scope": "User"

    ],
    "grant_types": [
    "authorization_code",
    ],
    "client_secret_expires_at": 0,

>     "scope": "openid",

Reviewer: same here 
"scope": "User"
